lib/puppetx/augeasproviders_grub/util.rb

```diff
@@ -152,7 +152,9 @@ def self.grub2_cfg_paths
         valid_paths = paths.map do |path|
           begin
             real_path = File.realpath(path)
-            real_path if (File.readable?(real_path) && !File.directory?(real_path))
+            real_path if (File.readable?(real_path) && !File.directory?(real_path)) &&
+               # Exclude stub files which include main config (e.g. Debian OS family)
+               (File.foreach(real_path).inject(0) {|c, line| c+1} > 5)
```

Comment:
Is there any way to authoritatively determine whether or not these files should be included?

The proposed method seems like it leaves too much up to chance.

Reply:
The authoritative way of the distributions in question is to never touch any of the files, but instead always call `update-grub`, which only touches `/boot/grub/grub.cfg`. This is essentially the reason why https://github.com/hercules-team/augeasproviders_grub/pull/57 broke all Debian systems. 

So I believe the only other option we have would be to hardcode a list of distributions: Some for which we are only allowed to touch this single file, and other distributions for which we can / should touch all files. 

The problem is that this may break when distributions change, and for another distribution, `/boot/grub/grub.cfg` may not be the authoritative file. So this may end up to become a maintenance hell. 

So I think we only have two options here:
* Identify stub files somehow. This may be done by length, or content — but the content is not straightforward to parse, e.g. on Debian / Ubuntu we have:
```
search.fs_uuid ed4b34fd-eac3-4c8b-9afb-07a92f5f88b5 root hd0,msdos3 
set prefix=($root)'/boot/grub'
configfile $prefix/grub.cfg
```
Checking for a line in which another config file is included may increase stability slightly. 
* Hardcoding a list for all distributions and their versions (past and future). 

Any better idea is very welcome.

Reply:
The original reasons for hitting all of the files were:
1. Red Hat would update the EFI or non-EFI file depending on how you were booted at the time
2. There wasn't a great way to tell if someone had mucked about with things by hand

I think that checking for `configfile` is a pretty solid check. I don't see that in any of my RHEL-based systems so it should have the desired effect.

Reply:
> The original reasons for hitting all of the files were: [...]

I understand — it's always astonishing how different distros take such different approaches ;-). 


> I think that checking for `configfile` is a pretty solid check. I don't see that in any of my RHEL-based systems so it should have the desired effect.

All the Debian systems I have access to have `configfile` in their stub configs. Do you think this check itself is sufficient, and I should drop the length check? 
I can implement and test this tomorrow (will then give it a run on our Ubuntu 18.04, Debian 10 and CentOS 7 / 8 nodes).

Reply:
It seems like `configfile` should be enough. I mean, the length could be any number of arbitrary things and we *do* want to update symlinks in case they start shoving things in other locations.

Reply:
>  I can implement and test this tomorrow (will then give it a run on our Ubuntu 18.04, Debian 10 and CentOS 7 / 8 nodes).

Note: I have already force-pushed the new implementation (well, it's a one-liner of course). I have for now only done basic testing on Debian 10, more to follow tomorrow :-).

Reply:
@trevor-vaughan I have concluded my tests on all systems I have access to:
* Ubuntu 18.04, booting via EFI or legacy / BIOS. 
* Debian 10, booting via EFI or legacy / BIOS. 
* CentOS 7 and 8, both booting via legacy / BIOS. 

The new approach works (expectedly) well ;-).